### PR TITLE
doc: various developer notes updates

### PR DESCRIPTION
**Pull request by jonatack**
A few updates noticed while working on a lifetimebound section.

- Remove obsolete BDB ENABLE_WALLET section (only one file, src/wallet/bdb.h, still has a `db_cxx.h` BDB header)
- Prefer Python for scripts in developer notes (and a few miscellaneous touch-ups)
- In the code style section, add a link to the internal interface exception so that people are aware of it
- Add a "C++ functions and methods" section
- Add a Clang `lifetimebound` attribute section

### CODE REVIEW

File: doc/developer-notes.md [hunk 1]
```diff
@@ -1390,7 +1391,7 @@ communication:
   ```
 
 - For consistency and friendliness to code generation tools, interface method
-  input and inout parameters should be ordered first and output parameters
+  input and in-out parameters should be ordered first and output parameters
   should come last.
```
Comment (laanwj) [inline doc/developer-notes.md:1395]: Maybe we can add something that we prefer returning values directly to "output parameters". This was written before the days of `std::optional` etc.

Comment (jonatack) [reply doc/developer-notes.md]: Good idea, added 968a67b9be5262228b47184b9821fcab as a first attempt.

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25092/commits/91d6d4a2337719cce48532dd614bb345be66fbbf

File: doc/developer-notes.md [hunk 2]
```diff
@@ -138,6 +141,16 @@ public:
 } // namespace foo
 ```
 
+Coding Style (C++ functions and methods)
+--------------------
+
+- Input and in-out parameters should be ordered first and output parameters
+  should come last.
```
Comment (laanwj) [inline doc/developer-notes.md:148]:
```
- *Rationale*: API consistency.
- ```

File: doc/developer-notes.md [hunk 3]
```diff
@@ -138,6 +141,16 @@ public:
...
+
+- Prefer returning values directly (using `std::optional`, where helpful) to
+  using output parameters.
```
Comment (laanwj) [inline doc/developer-notes.md:151]:
```
- *Rationale*: Less error-prone (no need for assumptions what output is initialized to on failure). Easier to read.
```

File: doc/developer-notes.md (hunk 2 above)
Comment (jonatack) [reply doc/developer-notes.md]: Added (thanks!)

File: doc/developer-notes.md (hunk 3 above)
Comment (jonatack) [reply doc/developer-notes.md]: Added (thanks!) along with some ideas from https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs.

File: doc/developer-notes.md [hunk 4]
```diff
@@ -138,6 +141,22 @@ public:
...
+Coding Style (C++ functions and methods)
+--------------------
+
+- When ordering function parameters, place input parameters first, then any
+  in-out parameters, followed by any output parameters.
+
+- *Rationale*: API consistency.
+
+- Prefer returning values directly to using in-out or output parameters. Use
+  `std::optional` where helpful for returning values (as well as for optional
```
Comment (laanwj) [inline doc/developer-notes.md:153]: Good point. I think using `std::optional` *for optional inputs* (instead some magic default value, if there is no real default) has enough merit to be called out separately.

Comment (jonatack) [reply doc/developer-notes.md:153]: > has enough merit to be called out separately.

Suggesting splitting it out to a separate bullet point?

Comment (laanwj) [reply doc/developer-notes.md:153]: I do think that's clearer.

Comment (jonatack) [reply doc/developer-notes.md]: Done.

**Review by jarolrod — COMMENTED** [review]
code review ACK https://github.com/bitcoin/bitcoin/commit/654284209f30fd309c326a527cda1d2d7385cee8

Nice improvements to the dev docs 👍

**Pull request merged**